File: leetcode/C++/longest-string-chain.cpp
class Solution {
public:
    bool checkPredecessor(string &s, string &ls){
        for(int i=0,j=0,d=0; i<s.length(); ++i,++j){
            if(s[i]!=ls[j]){
                --i;
                if(d++) return false;
            }
        }
        return true;
    } 
    int longestStrChain(vector<string>& words) {
        int sol=1;
        vector<int> acc(words.size(),1);
        sort(words.begin(),words.end(),[](const string &a, const string &b){return a.length()<b.length();});
        for(int i=0; i<words.size(); ++i){
            for(int j=i+1,t; j<words.size()&&words[j].length()<=words[i].length()+1; ++j){
                if(words[j].length()==words[i].length()) continue;
                if(acc[i]+1 > acc[j] && checkPredecessor(words[i],words[j])){
                    acc[j]=acc[i]+1;
                    if(sol<acc[j]) sol=acc[j];
                }
            }
        }
        return sol;
    }
};

File: leetcode/Python/longest-string-chain.py
class Solution:
    def checkPredecessor(self, s, sl):
        i,j,d=0,0,0
        while i<len(s):
            if s[i]!=sl[j]:
                if d>0:
                    return False
                i, d = i-1, d+1
            i,j = i+1, j+1
        return True
    def longestStrChain(self, words: List[str]) -> int:
        sol, acc = 1, [1 for i in range(len(words))]
        words.sort(key=len)
        for i in range(len(words)):
                for j in range(i+1,len(words)):
                    if len(words[i])+1<len(words[j]):
                        break
                    if len(words[i]) == len(words[j]):
                        continue
                    if acc[i]+1>acc[j] and self.checkPredecessor(words[i],words[j]):
                        acc[j]=acc[i]+1
                        sol=max(sol,acc[j])
        return sol